Update challenge format

escottalexander opened the pull request:
Experimenting with a different challenge format.
This PR:

moves all of the instruction to the README.
generalizes tests so that user has more leeway,
empties the contract file except for default import and pragma
adds reentrancy test to the test file

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,257 @@
+     * - Mark the address as a member.
+     * Emits a `MemberAdded` event.
+     */
+    function addMember(address newMember) external onlyOwner {
```
The DAO should get to decide if a member is added or removed. Anyone should be able to call these methods as long as a proposal to add/remove a member has been passed.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,257 @@
+     * - Revert with `InsufficientETH` if the amount exceeds the contract's balance.
+     * - Revert with `FailedTransfer` if the transfer fails.
+     */
+    function withdraw(uint256 amount) public onlyOwner {
```
I don't think this function has a purpose. The DAO is collectively owned so there is no single owner. You can remove the Ownable import as well.

escottalexander replied:
You can add a method that executes calldata defined in a proposal. This will be how the DAO's balance changes. When the DAO votes to send funds somewhere and it is approved, anyone can call the method to send those funds.

escottalexander commented on packages/foundry/contracts/MaliciousContract.sol:
Curious what you had in mind for this. It is out of scope for the challenge but pretty cool that you were thinking about this attack vector. We will probably utilizing reentrancy in our tests for most challenges in the future but this may be a little premature.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,257 @@
+     * - Revert with `FailedTransfer` if the transfer fails.
+     * Emits a `RageQuit` event.
+     */
+    function rageQuit() external onlyMember {
```
This looks great!

escottalexander reviewed (changes requested):
Gave it an initial check over. I outlined a few things that need to change but you are headed in the right direction overall.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
I am rethinking this.
1. Let's make the propose method accept a `contractAddr`, `data` and `value` param (see #33 for an example of this) and remove `ethAmount` and `shareAmount` from the struct. Also add a `deadline` param so that there is a time window for voting.
2. Add a new method called `executeProposal` or similar. Anyone can call it but it will check that the proposal has been approved and the proposal deadline is over. If true then it executes the data with the value. This way the method can be used to execute any transaction (not just approving members). If the proposal is rejected then it should refund any value that was deposited when the proposal was made.
3. Change the `addMember` method to have a modifier that only allows it to be called as the result of a proposal being executed (maybe `msg.sender == address(this)`?). `addMember` should accept all the params needed to give shares to the proposed new member.
4. `removeMember` should be very similar to add `addMember` but perform the opposite logic.

As you think through it you may find you have to add/remove other methods and logic that I haven't thought of yet.

KcPele replied:
i have updated it.

escottalexander replied:
Is there a reason why it is still a draft PR? I generally think of draft PRs as PRs that aren't ready for review.

escottalexander reviewed (changes requested):
I have some bigger thoughts for you that will require larger changes.

KcPele commented:
Great review

On Tue, 4 Jun 2024, 16:55 escottalexander, ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> I have some bigger thoughts for you that will require larger changes.
> ------------------------------
>
> On packages/foundry/contracts/MolochRageQuit.sol
> <https://github.com/BuidlGuidl/eth-tech-tree-challenges/pull/32#discussion_r1626256823>
> :
>
> I am rethinking this.
>
>    1. Let's make the propose method accept a contractAddr, data and value
>    param (see #33
>    <https://github.com/BuidlGuidl/eth-tech-tree-challenges/pull/33> for
>    an example of this) and remove ethAmount and shareAmount from the
>    struct. Also add a deadline param so that there is a time window for
>    voting.
>    2. Add a new method called executeProposal or similar. Anyone can call
>    it but it will check that the proposal has been approved and the proposal
>    deadline is over. If true then it executes the data with the value. This
>    way the method can be used to execute any transaction (not just approving
>    members). If the proposal is rejected then it should refund any value that
>    was deposited when the proposal was made.
>    3. Change the addMember method to have a modifier that only allows it
>    to be called as the result of a proposal being executed (maybe msg.sender
>    == address(this)?). addMember should accept all the params needed to
>    give shares to the proposed new member.
>    4. removeMember should be very similar to add addMember but perform
>    the opposite logic.
>
> As you think through it you may find you have to add/remove other methods
> and logic that I haven't thought of yet.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/BuidlGuidl/eth-tech-tree-challenges/pull/32#pullrequestreview-2096813495>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APJDTO4LVNSS3G6T4RBJJ4LZFXPQXAVCNFSM6AAAAABIQ6UQJOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAOJWHAYTGNBZGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

KcPele commented:
@escottalexander, I  want to get your feedback before I completely write out the test

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -149,6 +177,46 @@ contract MolochRageQuit {
+            proposal.data
+        );
+        if (!success) {
+            revert MolochRageQuit__FailedTransfer();
```
Maybe we need a new revert message that is failure to execute instead of failed transfer as that may not make sense if no value was being transferred.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -149,6 +177,46 @@ contract MolochRageQuit {
+            revert MolochRageQuit__ProposalNotApproved();
+        }
+
+        (bool success, ) = proposal.contractAddr.call{value: proposal.value}(
```
We should not send any value if the `proposal.contractAddr == address(this)` because the value doesn't need to move anywhere, it is already in the contract.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -161,17 +229,17 @@ contract MolochRageQuit {
     function exchangeShares(uint256 proposalId) external payable onlyMember {
```
Instead of onlyMember we should use onlyContractAddress on this method.

escottalexander replied:
It also shouldn't be external or payable. No value needs to move to use the function when executing a proposal.

escottalexander replied:
Also addMember's logic should be called  in this method OR we should define another method that handles calling both. We want a single proposal to be able to do both actions and the exchanging shares is part of becoming a member so no reason to separate them.

KcPele replied:
for  exchangeShares, it deals with msg.value that why it is payable and users send funds to the smart contract. if am to call exchangeShares in addMembers then I should also call rageQuit in remove member

```sol
function __exchangeShares(
        uint256 proposalId
    ) private payable onlyContractAddress {
        Proposal storage proposal = proposals[proposalId];
        if (proposal.proposer != msg.sender || !proposal.approved) {
            revert MolochRageQuit__ProposalNotApproved();
        }
        if (msg.value < proposal.value) {
            revert MolochRageQuit__InsufficientETH();
        }

        totalEth += msg.value;
        totalShares += proposal.value;
        shares[msg.sender] += proposal.value;

        emit SharesExchanged(msg.sender, msg.value, proposal.value);
    }
  function addMember(address newMember) external onlyContractAddress {
        if (members[newMember]) {
            revert MolochRageQuit__MemberExists();
        }
__exchangeShares()
        members[newMember] = true;
        emit MemberAdded(newMember);
    }


 function __rageQuit() private onlyContractAddress {}
  function removeMember(address member) external onlyContractAddress {
__rageQuite()
        members[member] = false;
        emit MemberRemoved(member);
    }
'''

escottalexander reviewed (changes requested):
Nice job on this. This is exactly what I had in mind. I made a few comments to further refine.

KcPele commented:
@escottalexander i have effected the changes

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,303 @@
+    event MemberRemoved(address member);
+    event Voted(uint256 proposalId, address voter);
+    event ProposalValueRefunded(address proposer, uint256 amount);
+    ///////////////////
```
Add a space above this line

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,303 @@
+        }
+        _;
+    }
+    ///////////////////
```
Add a space above this line

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,303 @@
+     * @param value The ETH value to send with the transaction.
+     * @param deadline The timestamp until when the proposal is valid.
+     * Requirements:
+     * - `contractAddr` should not be zero address.
```
We should tell the challenger what error to revert with in each of these cases.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,303 @@
+     * @dev Execute an approved proposal after the deadline.
+     * @param proposalId The ID of the proposal to execute.
+     * Requirements:
+     * - The proposal must be approved.
```
Tell the challenger what to revert with for each of these cases

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,303 @@
+    /**
+     * @dev Rage quit and exchange shares for ETH.
+     * Requirements:
+     * - The caller must have shares and must be a member.
```
Tell them what to revert with if it doesn't match this case

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,303 @@
+     * - Revert with ` MolochRageQuit__FailedTransfer` if the transfer fails.
+     * Emits a `RageQuit` event.
+     */
+    function __rageQuit(address member) private onlyContractAddress {
```
This can be called by any member. Not just the contract as the result of a proposal. Change the modifier.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,303 @@
+            revert MolochRageQuit__FailedTransfer();
+        }
+        emit RageQuit(member, memberShare);
+    }
```
It doesn't appear that the member is removed. I would expect `removeMember` to be called.

KcPele replied:
i made rageQuit handle everything

escottalexander reviewed (changes requested):
Left a few small comments. Getting really close with this one.

escottalexander commented:
@KcPele Can you resolve the merge conflicts please?

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0 <0.9.0;
+
+////////////////////
```
These errors should be defined inside the contract. I realize the layout guidelines are confusing as it lists contracts below errors but what this means is contracts variables defined inside the contract.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+    function propose(
+        address contractAddr,
+        bytes memory data,
+        uint256 value,
```
No need for this to be a parameter. We should just use `msg.value`.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+    // State Variables
+    ///////////////////
+    uint256 public totalShares;
+    uint256 public totalEth;
```
totalEth doesn't need to be defined or tracked in any way. You can get the contract balance with address(this).balance

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+////////////////////
+// Errors
+////////////////////
+error MolochRageQuit__NotAMamber();
```
~~NotAMamber~~ NotAMember

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+    uint256 public totalShares;
+    uint256 public totalEth;
+    uint256 public proposalCount;
+    uint256 public quorum;
```
The way quorum is used is somewhat flawed. It should probably be replaced with a memberCount and then a formula is used to make sure a majority has voted. It should never allow one person to be the sole decider of a vote unless there is only one member.

In your current setup the only place they can set the quorum is in the constructor which it could never be set to more than 1 or else there would be no way to add a second member. If the quorum remains one indefinitely then one vote decides the outcomes for any proposal to no matter how many members there are. Yikes.

Perhaps there should be a minimum quorum of voters (maybe as low as 5%) that is required for a proposal to be approved but there is a more robust yea vs nay formula used after the deadline has passed. Because maybe only 10% of members voted and maybe that minimum quorum was reached but there were more nay votes than yea votes. I don't want to over complicate the contract so I am open to your ideas but the current mechanism is broken.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+            revert MolochRageQuit__AlreadyVoted();
+        }
+
+        proposal.votes++;
```
In regards to my suggestions for updating the `quorum` variable: Maybe the proposal should track yea votes and nay votes instead of just votes in general.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+
+        if (proposal.votes >= quorum) {
+            proposal.approved = true;
+            emit ProposalApproved(proposalId, msg.sender);
```
I don't think it makes sense for a proposal to be approved before the voting period is over. More nay votes might counteract the current approval. The approval/denial should be decided in the `executeProposal function`. That is the only place where it being approved/denied matters.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+     * - if the proposal is approved and the deadline has passed: execute the calldata with the value. Reverts with MolochRageQuit__FailedToExecute if the execution fails.
+     * - Emit a `ProposalExecuted` event if the proposal is executed.
+     */
+    function executeProposal(
```
This function is overly complicated. It might help the user to instead have a `executeApprovedProposal` function and a `refundDeniedProposal` function. Each function could check if Yea > Nay or the opposite before executing.

escottalexander replied:
There is nothing to keep me from executing a proposal multiple times. Yikes! Make sure you set the new  property before `call`ing external contracts because this is a reentrancy risk. Would be great to add tests around this too.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+        emit Voted(proposalId, msg.sender);
+
+        if (proposal.votes >= quorum) {
+            proposal.approved = true;
```
Remove `approved` on the proposal and instead track yea votes and nay votes. Approval depends on the proposal deadline being reached, total votes > quorum, and proposal.yeaVotes > proposal.nayVotes

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+    ///////////////////
+    // Modifiers
+    ///////////////////
+    modifier onlyMember(address _member) {
```
It seems that the onlyMember modifier could be simplified to use msg.sender inside the modifier instead of accepting it as a param.

escottalexander commented on packages/foundry/contracts/MolochRageQuit.sol:
```diff
@@ -0,0 +1,312 @@
+    {
+        Proposal storage proposal = proposals[proposalId];
+
+        if (members[newMember]) {
```
This check is the only thing keeping someone from executing the proposal multiple times. But what if I get in at a low share price and Rage Quit at a high share price? Then I can become a member again at my low share price by reexecuting my old proposal! I could do this over and over until the contract was drained.

escottalexander reviewed (changes requested):
Gave this a deeper look and found some issues that need to be sorted through.

escottalexander committed:
Update challenge format

escottalexander commented:
We changed the format of the challenge and added it to a new repo. I based the new branch off of your changes. Thanks for your contribution! https://github.com/BuidlGuidl/eth-tech-tree-challenges/tree/moloch-rage-quit
